### ISSUE TITLE
Don't use replication for first app load

### ISSUE THREAD

**Issue by estellecomment** [medic/medic-webapp#2284]
Using filtered replication for the first app load makes the server work like crazy (this happens : https://github.com/medic/medic-projects/issues/317)
We can safely load only 4-6 phones at a time, so branches need to do it right, and to coordinate with each other.
It also means first load is complicated enough that we do it only once, for the prod instance. If it was easy, we could have users test on a test instance, and then wipe data and work on a real instance (and thus stop polluting prod data!)

If we can use a view query, and start replication only after that, we could make it much easier.

**Comment by SCdF** [comment]
How would this work, practically? We filter because we only want certain docs to be downloaded? I feel like if we can do this ticket, we can thus remove filtered replication completely, right? Not sure how first app load is different from other situations.

**Comment by estellecomment** [comment]
Well, for the first load you want all the data, but after that you only want the changes, so you can't reuse the same query. You'll need the changes feed, and to filter it. So filtered replication.

In practice, something like : 
  - query for current seq number on server, save it
 - one giant view query, or maybe several smaller ones, to get your data
 - start continuous replication from saved seq number. (hopefully any changes that happened between previous seq number and now are ignored if you already have them, and added nicely if not. To be tested)

**Comment by SCdF** [comment]
> start continuous replication from saved seq number

I think if we can do this then this is potentially a winner. At least, a temporary winner (like the Erlang was). I'm unsure how you "trick" Pouch into the replication having had occurred when it technically hasn't: perhaps this is part of it's API and you already know how, perhaps it's a dirty hack?

This also works well with the idea of taking our filter and 'compiling' it into a view. The basic idea being, rewrite the filter so it emits the doc id for every user that this document could apply to. I'm unsure if it's actually compilable in this way though, which is something we'd have to look into.

**Comment by garethbowen** [comment]
This will definitely improve performance, but I have some concerns. Firstly it's basically duplicating the code from the filter to the view so it'd be easy for the two to get out of sync (a change made to one but not the other). Secondly there might be race conditions because the seq you get won't accurately reflect the state of your data. For example. this could lead to replication pulling down the same version of the document as you already have, which pouch may or may not be smart enough to reconcile.

**Comment by estellecomment** [comment]
> it'd be easy for the two to get out of sync

True. That's a concern. 
I wonder if we can reuse the same code, and kanso it all up in the right places, or maybe add a `grunt copy-my-files-around` step.
> this could lead to replication pulling down the same version of the document as you already have, which pouch may or may not be smart enough to reconcile

Yes, that's what I meant in my "hopefully" statement. If it's not smart enough, well. Stop and reconsider.

**Comment by garethbowen** [comment]
> I wonder if we can reuse the same code

One's in erlang and the other's javascript so it'd be difficult.

**Comment by SCdF** [comment]
You can write views in Erlang too, but regardless the concept of the two functions are not really the same:
 - One is a filter that takes a username against a request and returns true or false for a document
 - One is a view that for every document emits the name (or names) of the users who should get this document

So they are almost the opposite of each other in some ways. I guess you could do something sqrrrly and change the filter to wrap the view code and do an index of into the potential emits, so the code would be something like:

```js
filter = function(doc, req) {
  fnForListOfUsernamesFromDoc(doc).indexOf(req.userId) !== -1
}

view = function(doc) {
  fnForListOfUsernamesFromDoc(doc).map(function(username) {
    emit(username);
  };
}
```

But that would stop you from being able to short-circuit out etc and generally wouldn't be as efficient, which is what we're trying to get away from in the first place.

You could probably pass an allowed short circuit bool, or write the function that supports either returning a list or short circuiting when you find a certain user, but still it's added complexity that is perhaps not ideal.

_(Unless I'm misunderstanding things ^_^)_

**Comment by SCdF** [comment]
Here is a quick attempt I pushed before the cafe kicked me out: https://github.com/medic/medic-webapp/pull/2293

It actually seems to work, but I'm not confident enough in anything to merge it. No perf tests etc. You get the idea though.

**Comment by estellecomment** [comment]
Interesting!
Agree that complexifying and slowing down the erlang is a bummer.

Or how about this : you leave the doc_by_place filter as-is, and each client makes a query with an ad-hoc  map function, specific to the querying user.
```
var currentUser;
var map = function(doc) {
  // somehow the client code gets a hold of doc_by_place...
  if(doc_by_place(doc, currentUser) {
   emit(doc._id)
  }
}
pouch.query(map, {include_docs: true});
```
(the ad-hoc view vs. stored view is to avoid having to store one view per user for ever, because each edit would trigger reindexing, for nothing)

Downside : ad-hoc map function means live index-building, which is probably really slow. But it only happens once so that could be ok. 
Bad connection might make it difficult...
Upside : you don't touch the doc_by_place filter, you can't get out of sync with it, and no side-effects on the rest of the system.

**Comment by estellecomment** [comment]
Good news, PouchDB replication will ignore the docs that you fetched with the view query!
Provided they have the same rev in both dbs, which you can achieve by saving them with `{new_edits: false}`.
http://wiki.apache.org/couchdb/HTTP_Bulk_Document_API#Posting_Existing_Revisions

Example script: https://github.com/medic/medic-webapp/blob/develop/scripts/first_load_with_view.js
Example log : [first_load_with_view.txt](https://github.com/medic/medic-webapp/files/267840/first_load_with_view.txt)
Example log, with pouch debug info : [first_load_with_view_debug.txt](https://github.com/medic/medic-webapp/files/267844/first_load_with_view_debug.txt)
You can see the revs_diff query, and that the docs that are already fetched are not `missing`, and thus not fetched.
Note that in this case, I started the replication from seq 0.

Next : 
 - try getting the remote seq number ahead of the whole process and start the replication there, to fetch less useless stuff.
 - a whole bunch of deleted docs are fetched, can we do without?

**Comment by estellecomment** [comment]
Now with pre-fetched seq number! You avoid fetching all the deleted docs, since you don't get early changes.
Note that once a replication is done, the next one will start from where the first left off (last_seq is saved in a _local doc). So you only need to specify the seq for the first time you run replication. 
That means when you reload the page, the seq number is stored, we don't need to persist it.

[first_load_with_view_saved_seq_number_debug.txt](https://github.com/medic/medic-webapp/files/267918/first_load_with_view_saved_seq_number_debug.txt)

**Comment by estellecomment** [comment]
Stumbled onto this timing experiment from a while ago: https://github.com/medic/medic-experiments/tree/master/replication-performance-v0.6

Timing replication vs view query, for the first load. Obviously the view query was much faster. The more docs in the db, the faster it is comparatively.

**Comment by garethbowen** [comment]
This is probably resolved by #2286 which replaces the filter for all replication not just first load.

**Comment by garethbowen** [comment]
We've now moved off filtered replication for all replications so this is resolved.
